[1.20.2] Endec - Simplier Codec Alternative

Dragon-Seeker opened the pull request:
Endec is an implementation based on a [previous attempt](https://github.com/Dragon-Seeker/owo-lib/tree/1.20-Kodeck) and [Codec](https://github.com/gliscowo/codec/blob/main/lib/codec.dart) by @gliscowo. Such is an attempt at being a similar but different alternative Minecraft DFU system Codec.

Features:
- Support for Networking by default (ByteBuf implementations like PacketByteBuf)
- Less Functional Programming
- Interop with Minecraft Codec

Such still requires refinement as such still lacks some of the various DFU implementations like MapCodec and possibly changes to general implementations.

Dragon-Seeker committed:
Initial Codeck implementation

BasiqueEvangelist commented:
code eck

Dragon-Seeker committed:
Method refactors and fix errors

- Adjust "any" methods to better reflect where they are located
- Fix error within writeAny for json and nbt
- Implement readAny NBT
- Add AW for forwarding Dyn Ops
- Fix errors within testing command

AlphaMode commented:
Bode ck

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/Codeck.java:
```diff
@@ -0,0 +1,306 @@
+
+    //--
+
+    //Kinda mega cursed but...
+    static <e, T> Codeck<T> of(BiConsumer<Serializer<e>, T> encode, Function<Deserializer<e>, T> decode) {
+        return new Codeck<T>() {
+            @Override
+            public <E> void encode(Serializer<E> serializer, T value) {
+                encode.accept((Serializer<e>) serializer, value);
+            }
+
+            @Override
+            public <E> T decode(Deserializer<E> deserializer) {
+                return decode.apply((Deserializer<e>) deserializer);
+            }
+        };
+    }
```
This would be a bit cleaner if `encode` and `decode` were bespoke interfaces that had the proper generics in them

Dragon-Seeker replied:
I really would not want to split such into separate interfaces

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/Codeck.java:
```diff
@@ -0,0 +1,306 @@
+
+    Codeck<BitSet> BITSET = Codeck.LONG_ARRAY.then(BitSet::valueOf, BitSet::toLongArray);
+
+    Codeck<Text> TEXT = Codeck.STRING.then(Text.Serializer::fromJson, Text.Serializer::toJson);
```
Doesn't Minecraft's `Codecs.TEXT` work more like `Codeck.JSON_ELEMENT`?

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/impl/MapCodeck.java:
```diff
@@ -0,0 +1,57 @@
+        return new MapCodeck<>(codeck, s -> s, s -> s);
+    }
+
+    public <R> MapCodeck<R, V> keyThen(Function<K, R> toFunc, Function<R, K> fromFunc) {
```
Is this method in the Public API? IntelliJ might complain about using it outside of owo if the impl package is made internal

Dragon-Seeker replied:
Possible, such is still up in the air about how the folder structure should be done.

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/impl/nbt/NbtDeserializer.java:
```diff
@@ -0,0 +1,264 @@
+
+    protected final Deque<Supplier<NbtElement>> stack = new ArrayDeque<>();
+
+    private boolean unsafe = true;
```
This doesn't really make it unsafe, though. All it does is make it lax on format errors

Dragon-Seeker replied:
I would call such unsafe but I am up for a better name

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/Codeck.java:
```diff
@@ -0,0 +1,306 @@
+    Codeck<ItemStack> ITEM_STACK = COMPOUND
+            .then(ItemStack::fromNbt, stack -> stack.writeNbt(new NbtCompound()));
+
+    Codeck<UUID> UUID = StructCodeckBuilder.of(
+            StructField.of("mostSig", Codeck.LONG, java.util.UUID::getMostSignificantBits),
+            StructField.of("leastSig", Codeck.LONG, java.util.UUID::getLeastSignificantBits),
+            UUID::new
+    );
```
This seems really clunky for formats like JSON (and not standard for NBT).

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/Codeck.java:
```diff
@@ -0,0 +1,306 @@
+            UUID::new
+    );
+
+    Codeck<Date> DATE = Codeck.LONG.then(Date::new, Date::getTime);
```
Same as for UUID.

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/Codeck.java:
```diff
@@ -0,0 +1,306 @@
+                    }
+            );
+
+    Codeck<BlockPos> BLOCK_POS = Codeck.LONG.then(BlockPos::fromLong, BlockPos::asLong);
+    Codeck<ChunkPos> CHUNK_POS = Codeck.LONG.then(ChunkPos::new, ChunkPos::toLong);
```
Same as UUID.

BasiqueEvangelist reviewed (commented):
æ

Dragon-Seeker committed:
Implement More Targeted approach for doing different methods of sterilization using Attributes

- Add Variable Length Ints

Dragon-Seeker committed:
Implement Codeck within Owo Networking

- Adjust networking test
- Fix improper encode and decode for String
- Use VAR_INT codeck for enum

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/Codeck.java:
```diff
@@ -0,0 +1,348 @@
+
+                return byteBuf;
+            }, byteBuf -> {
+                var bytes = new byte[byteBuf.readerIndex()];
```
```suggestion
                var bytes = new byte[byteBuf.readableBytes()];
```

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/Codeck.java:
```diff
@@ -0,0 +1,348 @@
+                            StructField.of("z", StructCodeck.INT, BlockPos::getX),
+                            BlockPos::new
+                    ),
+                    SerializationAttribute.HUMAN_READABLE
```
Won't this result in hard to edit NBT?

Dragon-Seeker replied:
Maybe, I have no clue how such would be represented, to say the least. It was mostly just a guess

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/Deserializer.java:
```diff
@@ -0,0 +1,43 @@
+
+    Set<SerializationAttribute> attributes();
+
+    Deserializer<T> addAttribute(SerializationAttribute ...attributes);
```
Codecks shouldn't be able to add attributes to deserializers

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/Endable.java:
```diff
@@ -0,0 +1,11 @@
+package io.wispforest.owo.serialization;
+
+/**
+ * Interface that indicates that the object has an end operation to such
+ */
```
IMHO it would be nicer to just make an interface that extends AutoCloseable and just makes `close` not throw anything, kind of like [`InfallibleCloseable` in gadget](https://github.com/wisp-forest/gadget/blob/1.20.2/src/main/java/io/wispforest/gadget/util/InfallibleClosable.java).

Dragon-Seeker replied:
Possibly a good idea as such exists just as a way to prevent issues with calling close method and `end` is just used due to what glisco used within her version.

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/SelfDescribedSerializer.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    default <V> void writeOptional(Codeck<V> codeck, Optional<V> optional) {
+        optional.ifPresentOrElse(v -> codeck.encode(this, v), this::empty);
```
Don't `NbtEnd`s disappear after serialization, making deserialization use the default field value instead?

Dragon-Seeker replied:
Unknown, keeping such until fully confirm that it is not required as I don't want to just discard such just yet

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/impl/CooptCodeck.java:
```diff
@@ -0,0 +1,109 @@
+        }
+
+        return error
+                ? DataResult.error(exception != null ? exception::getMessage : () -> "The corresponding Format for the given DynamicOps could not be located!")
```
This shouldn't really be a problem, as if you get a funky DynamicOps you can always just NbtOps or JsonOps and then convert everything to the needed dynamic ops

Dragon-Seeker replied:
True but I rather have this error instead to properly map such due to how conversion between formats can cause a lose of information potentially

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/impl/ListCodeck.java:
```diff
@@ -0,0 +1,41 @@
+        this.codeck = codeck;
+    }
+
+    public ListCodeck<V> mapConstructor(Supplier<List<V>> listConstructor){
```
```suggestion
    public ListCodeck<V> listConstructor(Supplier<List<V>> listConstructor){
```

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/impl/SerializationAttribute.java:
```diff
@@ -0,0 +1,8 @@
+    SELF_DESCRIBING(),
+    BINARY(),
+    HUMAN_READABLE(),
+    COMPRESSED();
```
As in compressed with gzip/whatever? Why would a Codeck want to care about this?

Dragon-Seeker replied:
As in the the data should be as small as possible but still hold onto some form of structure as in not intended to be human readable but still can be read in a sense based on order or something

Dragon-Seeker committed:
Adjust Attributes use case within De-Serilizers

Dragon-Seeker committed:
Remove SelfDescribedSerilalizer and handle Format Specific Codecks differently

Dragon-Seeker committed:
Fix incorrect constructor calls

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/Codeck.java:
```diff
@@ -0,0 +1,264 @@
+                    SerializationAttribute.HUMAN_READABLE)
+            .orElse(Codeck.LONG.then(ChunkPos::new, ChunkPos::toLong));
+
+    Codeck<BitSet> BITSET = Codeck.LONG_ARRAY.then(BitSet::valueOf, BitSet::toLongArray);
```
won't this lose the precise length of the bitset?

Dragon-Seeker replied:
Unknown but such was copied from PacketBufSerilizer which calls `writeBitSet` and `readBitSet` using a long array within `PacketByteBuf`

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/SequenceDeserializer.java:
```diff
@@ -0,0 +1,10 @@
+    boolean hasNext();
+
+    E next();
+}
```
This interface should probably have a way to get the total amount of stuff in this collection, as it's possible to get on NBT, JSON and PacketByteBuf and makes decoding into a list/whatever much more efficient

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/impl/StructCodeckBuilder.java:
```diff
@@ -0,0 +1,571 @@
+
+public class StructCodeckBuilder<T> {
+
+    public static <S, F1> Codeck<S> of(StructField<S, F1> f1, Function<F1, S> constructor){
+        return new StructCodeck<S>() {
+            @Override
+            public void encode(StructSerializer struct, S value) {
+                struct.field(f1.name, f1.codec, f1.getter.apply(value));
+            }
+
+            @Override
+            public S decode(StructDeserializer struct) {
+                return constructor.apply(struct.field(f1.name, f1.codec, f1.defaultValue));
+            }
+        };
+    }
```
Is there any way to make a struct field required?

Dragon-Seeker replied:
Fixed

Dragon-Seeker committed:
Preallocate list and map with the given size before adding entries

- Fix list method for changing the list constructor

Co-Authored-By: BasiqueEvangelist <basiqueevangelist@yandex.ru>

Dragon-Seeker committed:
Adjust Test Command for Codeck and Replace PacketBufSeriaizer within OwoHandshake

Dragon-Seeker committed:
Resolve issue with inablity to have StructField be called with required method call on StructDeserializer

Co-Authored-By: BasiqueEvangelist <basiqueevangelist@yandex.ru>

Dragon-Seeker committed:
Rename unsafe to safeHandling

Dragon-Seeker committed:
Change name to Endec as the possible final name

Dragon-Seeker committed:
Fix Coopt Class name and add easy way of converting to DFU Codec

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/network/OwoHandshake.java:
```diff
@@ -28,16 +31,15 @@
     @SuppressWarnings("unchecked")
-    private static final PacketBufSerializer<Map<Identifier, Integer>> RESPONSE_SERIALIZER =
-            (PacketBufSerializer<Map<Identifier, Integer>>) (Object) PacketBufSerializer.createMapSerializer(Map.class, Identifier.class, Integer.class);
+    private static final Endec<Map<Identifier, Integer>> RESPONSE_SERIALIZER =
+            (Endec<Map<Identifier, Integer>>) (Object) ReflectionEndecBuilder.createMapSerializer(Map.class, Identifier.class, Integer.class);
```
This should probably be updated to use a proper MapEndec

Dragon-Seeker replied:
Possibly

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/Endec.java:
```diff
@@ -0,0 +1,269 @@
+
+    <E> T decode(Deserializer<E> deserializer);
+
+    default <E> E encode(Supplier<Serializer<E>> serializerCreator, T value){
```
why does this take a supplier if it just invokes it once anyways

Dragon-Seeker replied:
It is an easy way of constructing a supplier of the desired serializer type and returning the value held by such serializer as the return.

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/SequenceDeserializer.java:
```diff
@@ -0,0 +1,12 @@
+
+    int size();
+
+    boolean hasNext();
+
+    E next();
```
Doesn't Iterator have these methods anyways? 🤨

Dragon-Seeker replied:
Yes

Dragon-Seeker replied:
Did such as a note to what is being implemented exactly and needs to be implemented

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/impl/StructField.java:
```diff
@@ -0,0 +1,48 @@
+
+public class StructField<S, F> {
+
+    public MutableObject<F> defaultValue = null;
```
Why is this a MutableObject?

Dragon-Seeker replied:
Such acts as a way of telling if a given value is defaulted. Such allows for the default value of null as setting the mutable object to be a instance of what ever defaulted value wanted.

BasiqueEvangelist replied:
What
You mean like it can either be null (required), or non-null, containing the actual default value?

Dragon-Seeker replied:
Yes, with the default value allowing for either a value of the given type or null

BasiqueEvangelist reviewed (commented):
Ant & Dec

Dragon-Seeker committed:
Add test for CooptCodec and adjust how error recover is handled with deserialization

Dragon-Seeker committed:
Implement MapCarrier with Endec

- Add KeyField
- Adjust StructField to be a extenstion of KeyField
- Add various helper method ontop of KeyField and StructField
- Adjust Nbt and Json Serializers for Prefix support

Dragon-Seeker committed:
Implement BinaryNbtSerializer and fix class loading issue with UwUShapedRecipe

gliscowo committed:
[endec] initial cleanup and first refactors
 - remove compressed and binary attributes as they both serve little purpose
 - update SelfDescribedDeserializer#readAny to accept a Serializer as a visitor to construct data instead of converting to intermediate types
 - reimplement JsonEndec and NbtEndec on top of updated readAny(...) specification
 - experimental implementation for Endec.ofCodec

gliscowo committed:
[endec] move struct-, sequence- and map- (de)serializer types into base (de)serializer interfaces

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/serialization/impl/json/JsonDeserializer.java:
```diff
@@ -0,0 +1,291 @@
+            } else if (primitive.isBoolean()) {
+                visitor.writeBoolean(primitive.getAsBoolean());
+            } else {
+                var value = primitive.getAsBigDecimal();
+
+                try {
+                    var asLong = value.longValueExact();
+
+                    if ((byte) asLong == asLong) {
+                        visitor.writeByte(element.getAsByte());
+                    } else if ((short) asLong == asLong) {
+                        visitor.writeShort(element.getAsShort());
+                    } else if ((int) asLong == asLong) {
+                        visitor.writeInt(element.getAsInt());
+                    } else {
+                        visitor.writeLong(asLong);
+                    }
+                } catch (ArithmeticException bruh /* quite cringe java moment, why use an exception for this */) {
+                    var asDouble = value.doubleValue();
+
+                    if ((float) asDouble == asDouble) {
+                        visitor.writeFloat(element.getAsFloat());
+                    } else {
+                        visitor.writeDouble(asDouble);
+                    }
+                }
```
couldn't you just use `.getAsNumber` and then check the type of the returned value?

gliscowo committed:
[endec] clean up all (de)serializer implementations, standardize (de)-serializer method ordering, remove binary nbt format

Dragon-Seeker committed:
Just going to leave this here

This is for the few that have been left behind with Mojang's implementation of Codec and lack the ability to handle any of their 17 field objects.

gliscowo committed:
[endec] do you like electronic dance music?

gliscowo committed:
[endec] add support for assuming attributes when embedding endec into codec

gliscowo committed:
[endec] remove reimplementations of vanilla recipes codecs and associated accesswidener entries

gliscowo committed:
[endec] quote field names in deserializer error messages

gliscowo committed:
[endec] superficial cleanup

Dragon-Seeker committed:
Cleanup ReflectionEndecBuilder

Also add support for nested generic calls with Map, List and Set on the getGeneric method

gliscowo committed:
[endec] initial cleanup on endec interface
 - separate non-primitive endecs into BuiltInEndecs
 - use wildcard generics on (De)Serializer parameters
 - implement Endec.dispatchedStruct for DFU parity

gliscowo committed:
[endec] remove MapEndec and ListEndec classes, being adding basic documentation to Endec interface

gliscowo committed:
[endec] refactor MapCarrier implementations and related infrastructure (always require a default value), refactor StructField to be independent of now-different KeyedEndec

gliscowo committed:
[endec] add Endec.forEnum

gliscowo committed:
Merge branch '1.20.2' into 1.20.2-Codeck
